aea/helpers/multiaddr/base.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+import base58
+
+import multihash
+
+from binascii import unhexlify
+
+from aea.crypto.base import Crypto
+
+from .crypto_pb2 import KeyType, PublicKey
+
+# NOTE: Implementation inspired from https://github.com/libp2p/py-libp2p
+# NOTE: On inlining...
+# See: https://github.com/libp2p/specs/issues/138
+# NOTE: enabling to be interoperable w/ the Go implementation
+
+ENABLE_INLINING = True
+MAX_INLINE_KEY_LENGTH = 42
+IDENTITY_MULTIHASH_CODE = 0x00
+
+KEY_SIZE = 32
+ZERO = b"\x00"
+
+if ENABLE_INLINING:
+
+    class IdentityHash:
+        _digest: bytes
+
+        def __init__(self) -> None:
+            self._digest = bytearray()
+
+        def update(self, input: bytes) -> None:
+            self._digest += input
+
+        def digest(self) -> bytes:
+            return self._digest
+
+    multihash.FuncReg.register(
+        IDENTITY_MULTIHASH_CODE, "identity", hash_new=lambda: IdentityHash()
+    )
+
+
+def _pad_scalar(scalar):
+    return (ZERO * (KEY_SIZE - len(scalar))) + scalar
+
+
+def _pad_hex(hexed):
+    # Pad odd-length hex strings.
+    return hexed if not len(hexed) & 1 else "0" + hexed
+
+
+def _hex_to_bytes(hexed):
+    return _pad_scalar(unhexlify(_pad_hex(hexed)))
+
+
+class MultiAddr:
+    def __init__(self, host: str, port: int, key: Crypto):
```

Comment:
We should pass the public key rather than the whole Crypto object. That makes the MultiAddr class usable in settings where all you have is the public key (and port and host) of some other agent, say.

Reply:
Right. I made Crypto object as part of the API to ensure users will pass the correct format of the public key.
I'll change it to a str object, with a TODO comment to add correct format checks